cmd/gcp-controller-manager/dpwi/serviceaccounts/verifier.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package serviceaccounts provides a common verifier to verify if a Kubernetes Service Account (KSA)
+// can act as a GCP Service Account (GSA). It also listens to and process KSA events.
+// If a KSA's permission changes, it notifies the configmap handler to update configmap
+// and the node syncer to sync related nodes.
+package serviceaccounts
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+	core "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/cloud-provider-gcp/cmd/gcp-controller-manager/dpwi/hms"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// serviceAccountAnnotationGSAEmail is the key to the GCP Service Account annotation
+	// in ServiceAccount objects.
+	serviceAccountAnnotationGSAEmail = "iam.gke.io/gcp-service-account"
+)
+
+// Verifier verifies if a Kubernetes Service Account (KSA)
+// can act as a GCP Service Account (GSA) or not.
+type Verifier struct {
+	hms         *hms.Client
+	saIndexer   cache.Indexer
+	verifiedSAs *saMap
+	loadGroup   singleflight.Group
+}
+
+// NewVerifier creates a new Verifier.
+func NewVerifier(saInformer coreinformers.ServiceAccountInformer, hmsAuthzURL string) (*Verifier, error) {
+	hms, err := hms.NewClient(hmsAuthzURL, &clientcmdapi.AuthProviderConfig{Name: "gcp"})
+	if err != nil {
+		return nil, err
+	}
+	v := &Verifier{
+		hms:         hms,
+		saIndexer:   saInformer.Informer().GetIndexer(),
+		verifiedSAs: newSAMap(),
+	}
+	return v, nil
+}
+
+// GSAEmail identifies a GCP service account in email format.
+type GSAEmail string
+
+// ServiceAccount identifies a K8s service account object by its namespace and name.  Empty
+// Namespace indicates the corresponding Kubernetes object was created in the "default" namespace.
+type ServiceAccount struct {
+	Namespace, Name string
+}
+
+// MarshalText implements the encoding.TextMarshaler interface.
+func (sa ServiceAccount) MarshalText() ([]byte, error) {
+	return []byte(sa.String()), nil
+}
+
+// String returns sa in a string as "<namespace>/<name>" or "default/<name>" if sa.Namespace is
+// empty.
+func (sa ServiceAccount) String() string {
+	if sa.Namespace == "" {
```

Comment:
I don't think this should be necessary.  Can this actually happen?

Reply:
It can be "" or "default". A customer found the  issue that we didn't handle it correctly.